hail/src/main/scala/is/hail/expr/ir/streams/StagedMinHeap.scala

```diff
@@ -0,0 +1,264 @@
+
+          val child = cb.newLocal[Int]("child")
+          cb.loop { Lrecur =>
+            // left child
+            cb.assign(child, index * 2 + 1)
+            cb.if_(child < heap.size,
+              cb.if_(
+                cb.invokeCode[Int](compareAtIndex, cb.this_, child, index) < 0,
+                cb.assign(smallest, child)
+              )
+            )
+
+            // right child
+            cb.assign(child, index * 2 + 2)
+            cb.if_(child < heap.size,
+              cb.if_(
+                cb.invokeCode[Int](compareAtIndex, cb.this_, child, smallest) < 0,
+                cb.assign(smallest, child)
+              )
+            )
+
+            cb.if_(smallest ceq index, cb.goto(Ldone))
+
+            cb.invokeVoid(swap, cb.this_, index, smallest)
+            cb.assign(index, smallest)
+            cb.goto(Lrecur)
+          }
+
+          cb.define(Ldone)
+        }
+      }
+
+    val pop_ : EmitMethodBuilder[_] =
+      classBuilder.defineEmitMethod("pop", FastSeq(), UnitInfo) { mb =>
+        mb.voidWithBuilder { cb =>
+          cb._assert(thisNonEmpty, s"${classBuilder.className}: poll empty")
+
+          cb.assign(garbage, garbage + 1L)
+          val newSize = cb.memoize(heap.size - 1)
+          cb.if_(newSize ceq 0, cb.assign(heap.size, newSize), {
+            cb.invokeVoid(swap, cb.this_, const(0), newSize)
+            cb.assign(heap.size, newSize)
+            cb.invokeVoid(heapify, cb.this_)
+          })
+        }
+      }
+
+    val append: EmitMethodBuilder[_] =
+      classBuilder.defineEmitMethod("append", FastSeq(elemParamType), UnitInfo) { mb =>
+        mb.voidWithBuilder { cb =>
+          heap.append(cb, mb.getSCodeParam(1))
+        }
+      }
+
+    val push_ : EmitMethodBuilder[_] =
+      classBuilder.defineEmitMethod("push", FastSeq(elemParamType), UnitInfo) { mb =>
+        mb.voidWithBuilder { cb =>
+          cb.invokeVoid(append, cb.this_, mb.getSCodeParam(1))
+          val Ldone = CodeLabel()
+          val current = cb.newLocal[Int]("index", heap.size - 1)
+          val parent = cb.newLocal[Int]("parent")
+
+          cb.while_(current > 0, {
+            cb.assign(parent, (current - 1) / 2)
+            val cmp = cb.invokeCode[Int](compareAtIndex, cb.this_, parent, current)
+            cb.if_(cmp <= 0, cb.goto(Ldone))
+
+            cb.invokeVoid(swap, cb.this_, parent, current)
+            cb.assign(current, parent)
+          })
+
+          cb.define(Ldone)
+        }
+      }
+
+    val arrayPType = PCanonicalArray(elemPType, required = true)
+
+    val toArray_ : EmitMethodBuilder[_] =
+      classBuilder.defineEmitMethod("toArray", FastSeq(typeInfo[Region]), arrayPType.sType.paramType) { mb =>
+        val region = mb.getCodeParam[Region](1)
+        mb.emitSCode { cb =>
+          arrayPType.constructFromElements(cb, region, heap.size, true) {
+            case (cb, idx) => heap.loadElement(cb, idx).toI(cb)
```

Comment:
I think you could use `arrayPType.sType.coerceOrCopy(cb, region, heap.eltArray.loadCheapSCode(cb, heap.data), true)` for this. That should be preferred as the canonical way to copy a value from one representation to another.

Reply:
This doesn't work: 
```
Caused by: is.hail.utils.HailException: tried to copy array with missing values to array of required elements
	at __C7MinHeapr_int32.toArray(/tmp/hail/__C7MinHeapr_int32.split.lir:814)
	at __C5Heapify.toArray(/tmp/hail/__C5Heapify.split.lir:68)
	at is.hail.expr.ir.StagedMinHeapSuite.$anonfun$heapify$2(StagedMinHeapSuite.scala:111)
	at is.hail.utils.package$.using(package.scala:657)
	at is.hail.annotations.RegionPool.scopedRegion(RegionPool.scala:162)
	at is.hail.expr.ir.StagedMinHeapSuite.$anonfun$heapify$1(StagedMinHeapSuite.scala:109)
	at is.hail.utils.package$.using(package.scala:657)
	at is.hail.expr.ir.StagedMinHeapSuite.$anonfun$gen$14(StagedMinHeapSuite.scala:176)
	at is.hail.utils.package$.using(package.scala:657)
	at is.hail.annotations.RegionPool.scopedRegion(RegionPool.scala:162)
	at is.hail.expr.ir.StagedMinHeapSuite.gen(StagedMinHeapSuite.scala:170)
	at is.hail.expr.ir.StagedMinHeapSuite.heapify(StagedMinHeapSuite.scala:108)
	at is.hail.expr.ir.StagedMinHeapSuite.$anonfun$testHeapProperty$1(StagedMinHeapSuite.scala:60)
	at is.hail.expr.ir.StagedMinHeapSuite.$anonfun$testHeapProperty$1$adapted(StagedMinHeapSuite.scala:59)
	at is.hail.check.GenProp1.$anonfun$apply$4(Prop.scala:28)
	at scala.runtime.java8.JFunction0$mcZ$sp.apply(JFunction0$mcZ$sp.java:23)
	at scala.util.Try$.apply(Try.scala:213)
	at is.hail.check.GenProp1.$anonfun$apply$3(Prop.scala:28)
```

Reply:
Oh right, because the `heap.eltArray` has extra missing elements at the end. Never mind!

Reply:
Thank you for explaining!